text/0000-variant-types.md

```diff
@@ -0,0 +1,335 @@
+this must just be an ad-hoc check in the compiler.
+
+
+# Drawbacks
+
+The variant types proposal is a little bit hairy, in part due to trying to
+remain backwards compatible.
+
+One could argue that having both tagged and untagged enums in a language is
+confusing. However, I believe the guidance here can be very clear: only use
+`#[repr(union)]` for C/FFI interop. The fact that it is an attribute should make
+it an obvious second choice.
+
+
+# Alternatives
+
+An alternative to allowing variants as types is allowing sets of variants as
+types, a kind of refinement type. This set could have one member and then would
+be equivalent to variant types, or could have all variants as members, making it
+equivalent to the enum type. Although more powerful, this approach is more
+complex, and I do not believe the complexity is justified.
+
+
+## Unsafe enums
+
+See [RFC PR 724](https://github.com/rust-lang/rfcs/pull/724) and
+[internals dicussion](https://internals.rust-lang.org/t/pre-rfc-unsafe-enums-now-including-a-poll/2873).
+
+Uses `unsafe` rather than an attribute to indicate that an enum is untagged.
+Uses an unsafe, irrefutable pattern match (let syntax) to destructure the enum,
+giving access to its fields.
+
+Using variant types and unsafe casting as proposed here should be more ergonomic
+- it better isolates the operation which is unsafe (discriminating the enum),
+from the safe operations (operating on the fields themselves).
+
+## Union structs
+
+See [RFC PR 1444](https://github.com/rust-lang/rfcs/pull/1444).
+
+Annotates structs rather than enums. This has the advantage over RFC 724 that
+fields can be accessed directly which is an ergonomic improvement (also true
+with this proposal). However, since all field access must be unsafe, it still
+requires more unsafe code than you might want.
+
+My preference is for an enum approach (as oppossed to structs) since a union
+offers multiple choices of data, like an enum, rather than combining data
+together like a struct. That is, enums and unions both 'or' data together,
+whereas structs 'and' data together. (In C, structs and unions are syntactically
+similar, but semantically very different).
+
+Furthermore, by using enums we allow union variants to have more than one field.
+While this is strictly more powerful than is needed for C interop, it is useful
+in general. For example, when dealing with binary data, formats will often have
+fields which are a given size, but may contain data of different types, an
+untagged enum is perfect for this.
+
+
+# Unresolved questions
+
+There is some potential overlap with some parts of some proposals for efficient
+inheritance: if we allow nested enums, then there are many more possible types
+for a variant, and generally more complexity. If we allow data bounds (c.f.,
+trait bounds, e.g., a struct is a bound on any structs which inherit from it),
+then perhaps enum types should be considered bounds on their variant types.
```

Comment:
Perhaps we can retain only trait bounds and use e.g. `V: VariantOf<E>` - which is what I've used in [my version](https://gist.github.com/eddyb/6c58b033b7621f436f47) of the examples in @nikomatsakis' [blog post introducing many of these ideas](http://smallcultfollowing.com/babysteps/blog/2015/08/20/virtual-structs-part-3-bringing-enums-and-structs-together/).